### ISSUE TITLE
[Proposal] Change of Websocket implementation

### ISSUE THREAD

**Issue by DarkGhostHunter** [beyondcode/laravel-websockets#754]
Hi BeyondCode. Love your stuff.

Things I don't like is stuff with many issues. Seems that some upstream dependencies are creating a lot of havoc regarding the WebSocket Server.

There is a [pure PHP implementation of WebSockets](https://github.com/amphp/websocket-server) that uses concurrency, created by the guys of amphp, the same group who figured out Fibers. It _seems_ stable.

I may create a PR that exchanges the upstream dependencies for that, which may fix a lot of problems thanks to the bloated dependency list.

**Comment by myckhel** [comment]
What do you think your proposal will improve from this library?

I think this library's websocket server doesn't have much issues but just some of the business logics.

**Comment by DarkGhostHunter** [comment]
Less dependencies, less moving parts, less code to write and maintain, easier debugging, potentially Fiber ready.

**Comment by myckhel** [comment]
Sounds nice 👍🏽

**Comment by radudiaconu0** [comment]
i think it will be cooler a php swoole /laravel octane implementation  would be cooler :) swoole already have a websocket serve impementation. i would do it myself but i don't know how to replace reactphp/rachet with swoole ;) @rennokki

**Comment by radudiaconu0** [comment]
https://www.swoole.co.uk/docs/modules/swoole-websocket-server

**Comment by DarkGhostHunter** [comment]
I tried to replace this library WebSocket implementation with Amphp at first, but I just couldn't make in tame. And anything we do now will be moot since Fibers is coming at the end of the year probably Laravel Octane too. 

The problem of Amphp and ReactPHP is that their async implementation are different. Amphp is more approachable, but once inside their async context, you must use their API, so there are not interoperable.

Swoole and Roadrunner (Laravel Octane) should yield performance miles away from Fibers, specially when using Swoole Redis / Roadrunner KV to access to Redis.

There are two options to make this package better:
1. Just go for Amp, as it has less moving parts and offers mature packages for WebSockets (Server, Redis) out of the box.
2. Wait for [Roadrunner for WebSockets](https://spiral.dev/docs/broadcast-websockets) ([should be soon](https://github.com/spiral/roadrunner/issues/513)) to implement them, along [Swoole](https://www.swoole.co.uk/docs/modules/swoole-websocket-server) later, for Laravel Octane, which would be the preferred option and offer great performance.

I honestly prefer option 2, since v2.0 its being finished as we speak, because you don't have to use Amp/ReactPHP or tie the package to their API, but just use plain PHP. At works, make interfaces to allow some kind of interoperability.

**Comment by radudiaconu0** [comment]
@DarkGhostHunter Yeah i am too for the swoole/roadrunner implementation:) It could be the fastest websockets server for laravel with this option:) I would be happy to help you

**Comment by DarkGhostHunter** [comment]
We may need to wait until Roadrunner WS implementation becomes available. Swoole and Roadrunner are very different servers, if you check Laravel Octane and how it deals with both of them.

**Comment by radudiaconu0** [comment]
@DarkGhostHunter yeah. Understood. I like this project more than laravel-echo-server tbh. :)) That project is so unmaintained:(

**Comment by DarkGhostHunter** [comment]
I may be able to make something that works as a page replacement, but note that it's not my priority right now.

**Comment by radudiaconu0** [comment]
@DarkGhostHunter i would be happy to help :)

**Comment by DarkGhostHunter** [comment]
We could start by having a branch for Octane development and first try a Roadrunner build with a Hello world. @mpociot

**Comment by mpociot** [comment]
Getting started on a branch sounds good to me
Why would you prefer Roadrunner over Swoole here?

**Comment by radudiaconu0** [comment]
@mpociot maybe because swoole doesn't support Linux for now

**Comment by radudiaconu0** [comment]
@DarkGhostHunter but we have to wait t'ill roadrunner release websockets support for v2

**Comment by mpociot** [comment]
Ah i didn't know that. 
Well either way, I personally focus on pushing out 2.0 shortly and afterwards I'm ready to start looking into this 🙂

**Comment by myckhel** [comment]
> @mpociot maybe because swoole doesn't support Linux for now. Maybe we can check which server Laravel WebSocket to use based on octane config. I guess thats the best bet.

Did you mean to say swoole doesn't support windows?

> Operation system: Linux, FreeBSD or MacOS
Linux kernel version >= 2.3.32
PHP version >= 7.0.0
GCC version >= 4.8

https://www.swoole.co.uk/docs/get-started/prerequisites

**Comment by radudiaconu0** [comment]
@myckhel  yeah sorry my mistake. I wanted to say it doesn't support WIndows. @mpociot

**Comment by DarkGhostHunter** [comment]
Yeah.

By the way, in case you don't know, the way Roadrunner works is very different from Swoole. 

Roadrunner acts more like a NGINX replacement. It doesn't exposes too much, but most of the heavy lifting is done by RR itself. With Swoole, more work has to be made, because it brings something like NGINX inside PHP, so you're responsable of dealing with the server in your PHP process rather than just exposing PHP-FPM to whatever.

Considering the convenience of RR implementation, it may do everything this package needs without too much code... or it will be weirdly inflexible to drop it. From what I saw, config is [absurdly simple](https://github.com/spiral/roadrunner/pull/660/files#diff-92b7146cec1e645d08829cb767ac102bab33df084337dec2bda6cefce4cb0315).

![image](https://user-images.githubusercontent.com/5141911/118920156-d1278280-b903-11eb-86db-f90860784f75.png)

In any case, [WS for RR](https://github.com/spiral/roadrunner/pull/660) is being pushed a lot, so I don't expect it to be ready soon.

**Comment by wolfy-j** [comment]
Just to be clear, RR will focus on WebSocket broadcast with memory-optimized BST on Golang end. While you **can** handle and serve WebSocket connections via PHP (Amp, Swoole) the routing and infrastructure code required to manage a large pool of connections will consume too much memory (that's PHP, you know). 

We are aiming to handle the massive amount of connections (via spatial and computation optimizations) on a single PHP node with the ability to scale horizontally using intermediate broadcast drivers.

So in short, WebSockets in Amp/Swoole will give you more low-level and more flexible access to the underlying connection at the cost of more memory. RoadRunner will hide a lot of WS abstractions in order to provide a ready-to-scale solution for WS data broadcasting (which in our opinion are 99% of WS use cases) out of the box.

**Comment by radudiaconu0** [comment]
@wolfy-j I got your point :) I still I think it would be better (with a lot more work) to change between RR and swoole based on Octane configuration. I mean one or another... But that's your decision not mine :P . In any case I am always here to help on both(swoole/RR ) implementations ;)

**Comment by wolfy-j** [comment]
@radudiaconu0 well, I was just trying to explain the difference, not vote against the implementation :)

**Comment by radudiaconu0** [comment]
@wolfy-j ah..ok Btw i love this package ;)

**Comment by DarkGhostHunter** [comment]
That's one of the reasons I'm waiting for RR-WS. Swoole would come later as I suspect many other apps also depend on Swoole on it and installing RR on top wouldn't make too much sense if memory is not a big problem.

**Comment by radudiaconu0** [comment]
sounds like a good idea:)

**Comment by radudiaconu0** [comment]
I see that RR websockets have been pushed yesterday :) @mpociot @DarkGhostHunter

**Comment by DarkGhostHunter** [comment]
Link or didn’t happen. 

Italo Baeza C.

**Comment by radudiaconu0** [comment]
https://github.com/spiral/roadrunner/pull/660

**Comment by DarkGhostHunter** [comment]
We're going to need docs, as I can't tell how to use it, let alone implementing channel auth for private/presence.

**Comment by wolfy-j** [comment]
@radudiaconu0 @DarkGhostHunter we did not complete the PHP SDK and Docs yet. It's not released yet. :)

**Comment by DarkGhostHunter** [comment]
Thanks, I'll stay on the look out.

**Comment by myckhel** [comment]
@mpociot @rennokki
Is this library abandoned?

Since no comments, moderation or any form of interactions from this library's authors for a long time.

**Comment by DarkGhostHunter** [comment]
1 month hardly can be defined as abandoned.

**Comment by DarkGhostHunter** [comment]
By the way, I'm currently waiting for proper docs for the Broadcasting capabilities of RR 2. I tried to make it work to no avail this weekend. cc: @wolfy-j

**Comment by rustatian** [comment]
Hey @DarkGhostHunter What problems did you face during implementation (what part of the docs do you think we are missed). Also, you may have a look at https://github.com/spiral/roadrunner-broadcast readme page, there are some docs and samples.

**Comment by DarkGhostHunter** [comment]
I tried to run it but I think I had something I got wrong on docker and WSL2. I'll try again later once my window opens up a little for a a proof of concept (the idea is to emulate pusher, which means authenticating channels).

**Comment by DarkGhostHunter** [comment]
Due to the lack of free time I'll have to sit this one out. Is still on my todo list but no ETA whatsoever. 

It should be possible to use Roadrunner WS, as it would (at a glance) offer superior performance by handling Redis and WS on their own bus rather than using PHP entirely. The only drawback is, well, Roadrunner as a dependency (hence, Octane), but shouldn't be a problem if the developer is already using this. This is my preferred approach.

A second alternative would be to require Revolt and build from that, which next versions of amphp and ReactPHP will support. This would be cleaner, but Redis would still have to be handled through an async library, which would defeat the purpose of only using Revolt.
